### DISCUSSION TITLE
Single Node Skeleton Poor Inference: Top-Down Not Great, and Bottom-Up Errors

### DISCUSSION THREAD

**Discussion by lisadiez** [talmolab/sleap#1170] [Help!]
Hello!

We are using Sleap to track gerbil huddles in our project (with a single node in the middle of the huddle). 

We labelled 600 frames (with many different huddle configurations and locations - see below) and trained a Top-down model. 
<img width="761" alt="training_images" src="https://user-images.githubusercontent.com/116014547/217337595-27271026-b5a9-4b15-b4d1-d7788fba3551.png">

But... The quality of the predictions is not great. 
We noticed that Sleap does not detect huddles for many, many frames at a time (even after it detected an exact same huddle one frame before):
<img width="1032" alt="prediction_error" src="https://user-images.githubusercontent.com/116014547/217338056-4374a9cb-bb8c-4455-8726-154d8d892323.png">

We're only getting predictions for 37% of the frames in this video, but the gerbils are always huddling (so it should actually be close to 100%):
<img width="893" alt="bar" src="https://user-images.githubusercontent.com/116014547/217338635-d591a825-82db-4941-9486-891f00e8e646.png">

We need to use a multi-animal model because sometimes we have more than one huddle in a single frame. 
We also tried using the Bottom-up model, but get an error when we try to train:
<img width="1312" alt="training_error" src="https://user-images.githubusercontent.com/116014547/217339009-2904d9c2-0044-4038-b4ca-d703083f6a73.png">

Any suggestions?
Should we add more frames or maybe tweak training/inference parameters?

Thank you!
Lisa

**Comment by roomrys** [comment]
Hi @lisadiez,

### Something to try:
I wonder if this error might be due to the crop size. Usually crop size for top-down is determined by the max distance between labeled points for a single instance. Since you only have one node, perhaps the crop is super(!) tiny. Could you try retraining by setting the crop-size to the approximate size of the largest huddle in pixels (I believe the max size we allow is 512px). You might need to eye-ball this huddle size given you know the image height/width.
 
### Questions:
I have a few follow-up questions (numbered below) to help determine both the cause of the bottom-up error and the poor inference w/ top-down:

1. Can you share your top-down model settings (the training config JSON would be perfect)
2. Can you share your bottom-up model settings

I'd like to troubleshoot your bottom-up problem first-hand, but it might be difficult to send over all the videos. When running via bottom-up, the error seems to come from SLEAP having trouble reading either the height or the width of one of your 9 videos and instead defaults to returning `None`.
https://github.com/talmolab/sleap/blob/833c2d5bdcf4cfef4e0adc7569b8e2245494a8fa/sleap/io/video.py#L1043-L1051

3. Which extension do your videos have? I am trying to determine which of SLEAP's video backends you are using (`HDF5Video`, `NumpyVideo`, `MediaVideo`, or `ImgStoreVideo`). Sending over just the slp to [lmaree@salk.edu](mailto:lmaree@salk.edu) is enough for me determine this if you'd rather do that.

It is strange that you only run into this with bottom-up since both bottom-up and top-down models should run through this section of code though (if `resize_and_pad_to_target` is set to `True` in the training config json).

Thanks,
Liezl
